Knative Project Security Self-Assessment - Security Pals

mayhumst opened the pull request:
## What?
We are a team of four students at New York University. We have created an assessment for the knative project based on cncf's Security Self-Assessment Template and would like to put it forward for review. We are in contact with the security team at knative and are continuing to make edits to the assessment. (Team members: @Jabberwockiii @mushr00m-blue @ShambhaviSeth )
## Why?
This assessment was the product of an assignment for a class with professor Justin Cappos, but we noticed that knative had no current assessment in the tag-security repository and would like to help fill this gap and make some security recommendations. 
## How?
This includes a current SBOM, a description of the product and its components/features, security goals and features, the deployment process, and recommendations for the documentation, among other sections. 
## Anything Else?
As previously mentioned, this assessment is a work in progress and all feedback is appreciated. We also hope to examine your threat model further and offer a few suggestions for revision in the next few days. 

Additionally, I am cc'ing @evankanderson , with whom our team has been in contact with and has provided excellent initial feedback.

eddie-knight commented:
Hi there! I'm just getting started looking at your pull request, and I noticed the DCO check is failing.

You can look at the checks section of the PR (I believe it should always be below the last comment) and look for a red X highlighting the failed check. In this case, you can click `Details` for more information about how to get that check passing.
<img width="315" alt="Screenshot 2023-12-08 at 8 35 18 AM" src="https://github.com/cncf/tag-security/assets/21176439/501ae793-0173-4e38-8ea2-57dca65ab1ea">

eddie-knight commented:
I noticed that you included an SBOM along with the self assessment. There are two reasons that jump to the front of my mind for why this isn't needed...

1. SBOMs should be associated with releases, as the bill of materials is only accurate and useful if it is created at build time and associated to a particular point in the code history.
1. If you need to link to an SBOM for some reason in the self-assessment, you can just provide a link out to the latest build artifacts that contain an SBOM.

We still have plenty more to review, but as a starter— could you please remove the SBOM from this PR?

mayhumst commented:
Thank you @eddie-knight for the initial feedback! I have removed the SBOM from the project folder, and I hope the go.mod files are enough for the SBOM section in the metadata. I squashed the commits and signed them off, so the DCO check should be good now.

ragashreeshekar commented on assessments/projects/knative/self-assessment.md:
```diff
@@ -0,0 +1,577 @@
+
+**Serving Controller** (Service) (https://knative.dev/docs/serving/): This is a central component in Knative responsible for orchestrating the deployment and scaling of serverless workloads. It operates independently, managing the lifecycle of Knative services and their revisions. The Serving Controller is isolated in such a way that even if it encounters issues or vulnerabilities, these do not d ...
+
+（suggestions from the maintainers: Knative Serving controllers and webhooks -- these manage most of the fields in the custom resources, including realizing lower-level Kubernetes resources (e.g. Deployments, k8s Services, certificate requests, etc).  These drive higher-level intents to lower-level implementations.  These implement all the public Knative Serving resources (Service, Route, Configur ...
```
Can you please clarify what does "suggestions from the maintainers" mean? 
What and Who are the suggestions for? 
Is it additional information maintainers would like to convey?

Jabberwockiii replied:
Thank you for your feedbacks! Those are the suggestions we got from the maintainer Evan Anderson. I included here as a skeleton for the actors part as people might have different views on which part is qualified as actors. Let me know if you want me to take that off.

ragashreeshekar commented on assessments/projects/knative/self-assessment.md:
```diff
@@ -0,0 +1,577 @@
+
+Knative emerged as a solution to these challenges. It was designed to simplify the process of deploying and managing serverless applications on Kubernetes. By providing middleware components, Knative enables developers to focus more on building their applications rather than worrying about the underlying infrastructure.
+
+### Actors
```
This section is very detailed with several actors, and it's easy to get lost in it. 
Can you please focus add an architectural diagram which helps relate different components in the system, and limit the description to the knative actors ("individual parts of your system that interact to provide the desired functionality.") . In addition, actors only need to be separate, if they are isolated in some way.

Jabberwockiii replied:
Do you think if we should include Functions in our actors?

evankanderson replied:
Yes, particularly since many of the components of functions run on the user's laptop, vs running on the cluster like the other components.

ragashreeshekar commented on assessments/projects/knative/self-assessment.md:
```diff
@@ -0,0 +1,577 @@
+7. Container-Based Function Execution: Functions in Knative are executed in containers, leveraging Kubernetes' container orchestration capabilities. This approach ensures portability, scalability, and consistent execution environments for functions.
+Further detailed Eventing components
+
+### Actions
```
The intent of this section is to showcase what each of the project performs in order to provide some service or functionality, however what I see here is what functionalities knative has and how to use knative. Please update this section to suit actions per our template - https://github.com/cncf/tag-security/blob/main/assessments/guide/self-assessment.md#actions.

A good reference is https://github.com/cncf/tag-security/blob/main/assessments/projects/pixie/self-assessment.md#actions

ragashreeshekar commented on assessments/projects/knative/self-assessment.md:
```diff
@@ -0,0 +1,577 @@
+        - Rollback : Ability to roll back to a previous version of the function if needed.
+
+
+### Goals
```
This section intends to cover key goals (typically 3-4) of the project and is not limited to security goals only. Please elaborate on the goals of the project

evankanderson replied:
* Minimize the amount of application developer overhead to build and deploy stateless applications
* Offer a streamlined developer experience for application developers 
* Provide an implementation of serverless compute and event handling infrastructure on Kubernetes

ragashreeshekar commented on assessments/projects/knative/self-assessment.md:
```diff
@@ -0,0 +1,577 @@
+| Component | Applicability | Description of Importance |
+| --------- | ------------- | ------------------------- |
+| Code Signature Verification | Security Relevant | Knative signs their releases with cosign to verify binaries, ensuring integrity and authenticity for users. |
+| Istio | Security Relevant | Istio is an open platform-independent service mesh that provides traffic management, policy enforcement, and telemetry collection. Although Istio is not required for Knative to operate, it can be used for networking operations like traffic rules, authorization policies, and encryption. |
```
This section should include functionalities within knative that enable key security functions. This is not meant to include third party tools or functionalities (such as Istio or security gaurd, i'm unfamiliar with the latter). Please update this section to reflect Knatives security functions

ragashreeshekar commented on assessments/projects/knative/self-assessment.md:
This section is  to understand if knative is complaint to any compliance or regulatory requirements out of the box. 
I do not think the following statement is accurate and applicable in this case. I suggest including references of knative project compliance. If found none, please record the same. 

> "However, Knative is built on Kubernetes, which follows the following regulatory standards depending on what industry it is deployed in: SOC 2, PCI DSS, HIPAA, NIST SP 800-53, GDPR, ISO 27001". 


Additionally,  the following statement is not relevant to this section. Please remove them. 

> "It could be said that Knative follows the CNCF Community Code of Conduct and the CNCF IP Policy. This should be easily accessible information and should be considered in the documentation".

ragashreeshekar commented on assessments/projects/knative/self-assessment.md:
Are there any additional security checks performed in this pipeline, like container image scan, SAST or secret scanning? Please include if any

ragashreeshekar commented on assessments/projects/knative/self-assessment.md:
Please group these communication channels into internal, inbound, outbound per the template - https://github.com/cncf/tag-security/blob/main/assessments/guide/self-assessment.md#secure-development-practices

ragashreeshekar commented on assessments/projects/knative/self-assessment.md:
Please move the recommendations to a new file in the same folder, and link them here.

ragashreeshekar reviewed (changes requested):
Thanks for the PR @mayhumst and team, appreciate the efforts.
I have completed first pass of review and left several comments on section that needs your attention. Please feel free to reach out here or on slack for any questions and clarifications.

Along with addressing the comments, kindly update the PR branch with the latest content in the repo as this branch is out-of-date with the base branch.

ragashreeshekar commented on assessments/projects/knative/self-assessment.md:
```diff
@@ -0,0 +1,577 @@
+| Software | https://github.com/knative/serving https://github.com/knative/eventing https://github.com/knative/func |
+| Security Provider? | No — the primary function of Knative is to provide a set of components and building blocks to extend Kubernetes functionality and should not be considered primarily a security provider. |
+| Languages | Go |
+| | |
```
Please include if there an SBOM published by the knative team. if there isn't one, please note the same

mayhumst commented:
@ragashreeshekar Thank you for detailed feedback! We've merged our branch to the parent repository, thank you for pointing that out. We've made a few initial changes based on your review and will fix the remaining issues quickly.

mayhumst commented:
Hi @ragashreeshekar Thank you again for the feedback. My teammate has updated the actors and actions sections as of this morning. I hope it's a better representation of this project. We welcome additional feedback and look forward to hearing back about next steps.

mayhumst reviewed (commented):
Two additional files have been added to the folder: one for general recommendations, and one image of knative's component diagram. 

We have updated the sections pointed out and hope the assessment makes better sense.

evankanderson commented on assessments/projects/knative/SBOM.json:
```diff
@@ -0,0 +1,218 @@
+
```
Oops, this was an earlier comment that hadn't been sent.  Ignore!

ragashreeshekar commented on assessments/projects/knative/self-assessment.md:
```diff
@@ -0,0 +1,545 @@
+| Software | https://github.com/knative/serving https://github.com/knative/eventing https://github.com/knative/func |
+| Security Provider? | No — the primary function of Knative is to provide a set of components and building blocks to extend Kubernetes functionality and should not be considered primarily a security provider. |
+| Languages | Go |
+| | |
```
```suggestion
| SBOM | Knative does not currently produce an SBOM at build time |
| | |
```

ragashreeshekar replied:
```suggestion
| SBOM | Unsure if SBOM is generated by the Knative team |
| | |
```
